feat: cursor sprites for the game editor tools

Arufonsu opened the pull request:
* Fixed issues when switching from marquee selected tiles to other tools.
* Cursor sprites are now switchable as the user switches between grid map tools.
* Moving the cursor out of the map-grid sets the cursor back to normal.
* Editor tool's custom sprites are configurable.
* Editor tool's custom sprite click-points are configurable.
* Should resolve #1453 

[assets pull request](https://github.com/AscensionGameDev/Intersect-Assets/pull/24)

### Preview: 

https://user-images.githubusercontent.com/17498701/191131619-5da88dbd-70f1-496c-ab5c-aa7919184d2e.mp4

Arufonsu commented:
_(Self Reminder for EnableEditorToolCursorSprites property)_

![image](https://user-images.githubusercontent.com/17498701/192125983-5e80695e-1a09-429d-b802-e60f17ee0df7.png)

![image](https://user-images.githubusercontent.com/17498701/192125978-f8a39334-560c-445d-910c-08262651478d.png)

Arufonsu commented:
> (Self Reminder for EnableEditorToolCursorSprites property)

Decided to just set the editor's EnableEditorToolCursorSprites to false
(Also fixed some issues that i've encountered when checking up some stuff).

Arufonsu commented:
Changed to draft for now, I'll use a struct for these for easier to read properties within the JSON file.

![image](https://user-images.githubusercontent.com/17498701/192160587-25e813a5-f64f-4eef-881b-9fd862e9df48.png)

Arufonsu commented:
Updated the branch, fixed merge conflicts, and refined the last (i hope) details for this new editor feature.

preview:
![image](https://user-images.githubusercontent.com/17498701/192167553-43427c28-2052-415a-b61e-105ad55633e9.png)

lodicolo commented on Intersect (Core)/Configuration/ClientConfiguration.cs:
```diff
@@ -148,9 +203,44 @@ public void Validate()
+        /// Sets a sprite (must end in .png) for the game client's cursor.
         /// </summary>
-        public string MouseCursor { get; set; } = "";
+        public string ClientCursorSprite { get; set; } = "";
```
Don't rename this, it's a breaking configuration change and this is going into a minor version

lodicolo commented on Intersect.Editor/General/Enums.cs:
```diff
@@ -48,15 +48,15 @@ public enum EditingTool
 
-        Rectangle = 2,
+        RectangleFill = 2,
 
-        Fill = 3,
+        LayerFill = 3,
 
-        Erase = 4,
+        Eraser = 4,
 
-        Droppler = 5,
+        Eyedropper = 5,
```
Do not rename the enum either

lodicolo replied:
That said I will add renaming some of these as part of #1169

lodicolo commented on Intersect.Editor/Localization/Strings.cs:
```diff
@@ -4049,7 +4049,7 @@ public partial struct Mapping
             public static LocalizedString eraselayerdialogue = @"Are you sure you want to erase this layer?";
 
-            public static LocalizedString filllayer = @"Fill Layer";
+            public static LocalizedString filllayer = @"Layer Fill";
```
Don't change this

lodicolo commented on Intersect.Editor/Localization/Strings.cs:
```diff
@@ -3815,7 +3815,7 @@ public partial struct MainForm
 
             [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
-            public static LocalizedString Fill = @"Fill Layer Tool (F)";
+            public static LocalizedString Fill = @"Layer Fill Tool (F)";
```
I don't think this string change makes sense, all of the tools (except the Eyedropper which conflicts with Eraser -- and Eyedropper has a more natural alternative with `i`) have a keybind that is the first letter of their tool name (at least in English).

That said, I do support changing all of the tool strings -- specifically drop the word "Tool" from the tool strings

lodicolo commented on Intersect.Editor/Forms/frmMain.cs:
```diff
@@ -367,11 +367,11 @@ private void FrmMain_KeyDown(object sender, KeyEventArgs e)
                         break;
 
-                    case Keys.F: // Fill Layer Tool.
+                    case Keys.F: // Layer Fill Tool.
```
Since it's just a comment I don't really care but I would personally revert this

lodicolo commented on Intersect.Editor/Forms/DockingElements/frmMapEditor.cs:
```diff
@@ -40,13 +42,41 @@ public partial class FrmMapEditor : DockContent
-            Icon = System.Drawing.Icon.ExtractAssociatedIcon(System.Reflection.Assembly.GetExecutingAssembly().Location);
-            picMap.MouseLeave += (_sender, _args) => tooltipMapAttribute?.Hide();
+            Icon = Icon.ExtractAssociatedIcon(System.Reflection.Assembly.GetExecutingAssembly().Location);
+            picMap.MouseLeave += (sender, args) => tooltipMapAttribute?.Hide();
```
If you're going to change the parameter names, change them both to `_` -- the underscore was to denote them not being used.

lodicolo commented:
I realized this after sleeping on it but the enable icons should not be in the JSON configuration, it should be a registry preference for the editor controlled by a UI toggle in the editor options (yes I know, annoying winforms changes) -- otherwise, with updates enabled individual users won't be able to turn it off/on, it will effectively be server controlled -- and that's not acceptable for this.

Now regarding the actual mouse cursor definitions -- I think you should create a `resources/cursors` directory and load sprites with the name format `$"editor_{enumValue.ToString()}.png"` -- and then the hotspots should be defined in `$"editor_{enumValue.ToString()}.json"` -- they should not all be defined in the config.json, these are all asset attributes not a real configuration (they're analogous to `resources/gui/layouts/*.json` files which are metadata for the interfaces -- this would be metadata for cursors).

Arufonsu commented:
Updated with @lodicolo 's review, [also updated the assets PR](https://github.com/AscensionGameDev/Intersect-Assets/pull/24)


### Preview

https://user-images.githubusercontent.com/17498701/192421772-c642ab75-3f4a-4b3b-8848-95a94597fe63.mp4

lodicolo commented on Intersect.Editor/Forms/DockingElements/frmMapEditor.cs:
```diff
@@ -2343,6 +2366,115 @@ private void picMap_Paint(object sender, PaintEventArgs e)
+
+            picMap.Focus();
+
+            switch (Globals.CurrentTool)
+            {
+                case (int)EditingTool.Brush:
+                    mCurPath = $"resources/cursors/editor_{EditingTool.Brush.ToString().ToLowerInvariant()}.png";
+                    mCurClickPoint = ToolCursor.Brush.CursorClickPoint;
+                    break;
+
+                case (int)EditingTool.Selection:
+                    mCurPath = $"resources/cursors/editor_{EditingTool.Selection.ToString().ToLowerInvariant()}.png";
+                    mCurClickPoint = ToolCursor.Selection.CursorClickPoint;
+                    break;
+
+                case (int)EditingTool.Rectangle:
+                    mCurPath = $"resources/cursors/editor_{EditingTool.Rectangle.ToString().ToLowerInvariant()}.png";
+                    mCurClickPoint = ToolCursor.Rectangle.CursorClickPoint;
+                    break;
+
+                case (int)EditingTool.Fill:
+                    mCurPath = $"resources/cursors/editor_{EditingTool.Fill.ToString().ToLowerInvariant()}.png";
+                    mCurClickPoint = ToolCursor.Fill.CursorClickPoint;
+                    break;
+
+                case (int)EditingTool.Erase:
+                    mCurPath = $"resources/cursors/editor_{EditingTool.Erase.ToString().ToLowerInvariant()}.png";
+                    mCurClickPoint = ToolCursor.Erase.CursorClickPoint;
+                    break;
+
+                case (int)EditingTool.Dropper:
+                    mCurPath = $"resources/cursors/editor_{EditingTool.Dropper.ToString().ToLowerInvariant()}.png";
+                    mCurClickPoint = ToolCursor.Dropper.CursorClickPoint;
+                    break;
+
+                default:
+                    return;
+            }
```
I said this switch didn't need to exist anymore. Using a dictionary this can be reduced to 2 assignment lines

Arufonsu replied:
```
mCurPath = $"resources/cursors/editor_{Globals.CurrentTool.ToString().ToLowerInvariant()}.png";
mCurClickPoint = ToolCursor.ToolCursorDict[(EditingTool)Globals.CurrentTool].CursorClickPoint;
```

lodicolo commented on Intersect.Editor/Forms/DockingElements/frmMapEditor.cs:
```diff
@@ -2343,6 +2366,115 @@ private void picMap_Paint(object sender, PaintEventArgs e)
+
+            mCurSprite = new Bitmap(mCurPath);
+
+            if (mCurSprite == null)
+            {
+                return;
+            }
```
This case is not possible here, as a constructor _cannot_ return null and this is the only place this variable is assigned to.

lodicolo commented on Intersect.Editor/Forms/DockingElements/frmMapEditor.cs:
```diff
@@ -2343,6 +2366,115 @@ private void picMap_Paint(object sender, PaintEventArgs e)
+        {
+            var enableCursorSprites = Preferences.LoadPreference("EnableCursorSprites");
+
+            if (!(!string.IsNullOrEmpty(enableCursorSprites) && Convert.ToBoolean(enableCursorSprites)) ||
+                string.IsNullOrEmpty(mCurPath) ||
+                !File.Exists(mCurPath))
+            {
+                return;
+            }
```
this check is not necessary

lodicolo commented on Intersect.Editor/Forms/DockingElements/frmMapEditor.cs:
```diff
@@ -2343,6 +2366,115 @@ private void picMap_Paint(object sender, PaintEventArgs e)
+                return;
+            }
+
+            mCurSprite.Dispose();
+            DestroyIcon(Cursor.Handle);
+            Cursor = Cursors.Default;
```
```csharp
if (mCurSprite != default) {
    ... this block of code ...
}
```

Arufonsu commented on Intersect.Editor/Configuration/ToolCursor.cs:
```diff
@@ -0,0 +1,73 @@
+                $"resources/cursors/editor_{EditingTool.Selection.ToString().ToLowerInvariant()}.json");
+        }
+
+        private void Validate()
+        {
+            CursorClickPoint = new Point(
+                CursorClickPoint.X != default
+                    ? CursorClickPoint.X
+                    : 0,
+                CursorClickPoint.Y != default
+                    ? CursorClickPoint.Y
+                    : 0);
+        }
+
+        [OnDeserialized]
+        internal void OnDeserialized(StreamingContext context)
+        {
+            Validate();
+        }
+
+        public ToolCursor Load(string filePath, bool failQuietly = false)
+        {
+            return ConfigurationHelper.Load(this, filePath, failQuietly);
+        }
+
+        public ToolCursor Save(string filePath, bool failQuietly = false)
+        {
+            return ConfigurationHelper.Save(this, filePath, failQuietly);
+        }
```
```csharp
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;
using Intersect.Editor.General;

namespace Intersect.Editor.Configuration
{
    public sealed class ToolCursor
    {
        public Point CursorClickPoint { get; set; }

        private static readonly Dictionary<EditingTool, ToolCursor> _toolCursorDict;

        public static IReadOnlyDictionary<EditingTool, ToolCursor> ToolCursorDict => _toolCursorDict;

        static ToolCursor()
        {
            _toolCursorDict = new Dictionary<EditingTool, ToolCursor>
            {
                { EditingTool.Brush, new ToolCursor() },
                { EditingTool.Dropper, new ToolCursor() },
                { EditingTool.Erase, new ToolCursor() },
                { EditingTool.Fill, new ToolCursor() },
                { EditingTool.Rectangle, new ToolCursor() },
                { EditingTool.Selection, new ToolCursor() }
            };
        }

        public static void Load()
        {
            foreach (EditingTool tool in Enum.GetValues(typeof(EditingTool)))
            {
                string fileName = $"resources/cursors/editor_{tool.ToString().ToLowerInvariant()}.json";

                if (_toolCursorDict.TryGetValue(tool, out ToolCursor cursor))
                {
                    // Cursor was found in dictionary, deserialize from file
                    string json = File.ReadAllText(fileName);
                    cursor = JsonConvert.DeserializeObject<ToolCursor>(json);
                    _toolCursorDict[tool] = cursor;
                }
                else
                {
                    // Cursor was not found in dictionary, create new instance and serialize to file
                    ToolCursor newCursor = new ToolCursor();
                    string json = JsonConvert.SerializeObject(newCursor);
                    File.WriteAllText(fileName, json);
                    _toolCursorDict.Add(tool, newCursor);
                }
            }
        }
    }
}

```
